File: module_book/src/main/java/com/ebook/book/mvvm/model/BookListModel.java
package com.ebook.book.mvvm.model;

import android.app.Application;

import com.ebook.db.GreenDaoManager;
import com.ebook.db.entity.BookInfo;
import com.ebook.db.entity.BookInfoDao;
import com.ebook.db.entity.BookShelf;
import com.ebook.db.entity.BookShelfDao;
import com.ebook.db.entity.ChapterListDao;
import com.xrn1997.common.mvvm.model.BaseModel;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class BookListModel extends BaseModel {
    public BookListModel(Application application) {
        super(application);
    }

    public Observable<List<BookShelf>> getBookShelfList() {
        return Observable.create((ObservableOnSubscribe<List<BookShelf>>) e -> {
                    List<BookShelf> bookShelves = GreenDaoManager.getInstance().getmDaoSession().getBookShelfDao().queryBuilder().orderDesc(BookShelfDao.Properties.FinalDate).list();
                    for (int i = 0; i < bookShelves.size(); i++) {
                        List<BookInfo> temp = GreenDaoManager.getInstance().getmDaoSession().getBookInfoDao().queryBuilder().where(BookInfoDao.Properties.NoteUrl.eq(bookShelves.get(i).getNoteUrl())).limit(1).build().list();
                        if (temp != null && !temp.isEmpty()) {
                            BookInfo bookInfo = temp.get(0);
                            bookInfo.setChapterlist(GreenDaoManager.getInstance().getmDaoSession().getChapterListDao().queryBuilder().where(ChapterListDao.Properties.NoteUrl.eq(bookShelves.get(i).getNoteUrl())).orderAsc(ChapterListDao.Properties.DurChapterIndex).build().list());
                            bookShelves.get(i).setBookInfo(bookInfo);
                        } else {
                            GreenDaoManager.getInstance().getmDaoSession().getBookShelfDao().delete(bookShelves.get(i));
                            bookShelves.remove(i);
                            i--;
                        }
                    }
                    e.onNext(bookShelves);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: module_book/src/main/java/com/ebook/book/mvvm/model/BookListModel.kt
package com.ebook.book.mvvm.model

import android.app.Application
import android.util.Log
import com.ebook.db.ObjectBoxManager.bookShelfBox
import com.ebook.db.entity.BookShelf
import com.ebook.db.entity.BookShelf_
import com.xrn1997.common.mvvm.model.BaseModel
import io.objectbox.query.QueryBuilder
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BookListModel @Inject constructor(
    application: Application
) : BaseModel(application) {
    fun getBookShelfList(): Observable<List<BookShelf>> {
        return Observable.create { e: ObservableEmitter<List<BookShelf>> ->
            bookShelfBox.query()
                .order(BookShelf_.finalDate, QueryBuilder.DESCENDING)
                .build().use { query ->
                    val bookShelves = query.find()
                    val iterator =
                        bookShelves.iterator()
                    while (iterator.hasNext()) {
                        val bookShelf = iterator.next()
                        Log.e("ttt", "getBookShelfList: " + bookShelf.id)
                        val temp = bookShelf.bookInfo.target
                        if (temp == null) {
                            bookShelfBox.remove(bookShelf)
                            iterator.remove()
                        }
                    }
                    e.onNext(bookShelves)
                }
        }.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
    }

}
